### PR TITLE
`feature/M3` | Improvements and bugfixes for M3

### PR DESCRIPTION
**Pull request by rrojan**
### Changes

- [x] Deal with all improvement and bugfix tasks for `feature/M3` here

### CODE REVIEW

File: src/app/_fetchers/ValidateCountFetcher.tsx [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { apiUrl } from '@/config'
+
+interface ValidateCountFetchProps {
+  token: string
+}
+
+export const ValidateCountFetch = async ({ token }: ValidateCountFetchProps) => {
```
Comment (sazanrjb) [inline src/app/_fetchers/ValidateCountFetcher.tsx:7]: `ValidateNotificationCountFetch` ?

File: src/app/api/notification/notification.service.ts [hunk 2]
```diff
@@ -336,4 +347,16 @@ export class NotificationService extends BaseService {
...
   }
+
+  async getAllForTasks<T extends UserRole>(
+    tasks: Task[],
+  ): Promise<T extends UserRole.Client ? ClientNotification[] : InternalUserNotification[]> {
+    const taskIds = tasks.map((task) => task.id)
+
+    if (this.user.role === UserRole.Client) {
+      return (await this.db.clientNotification.findMany({ where: { taskId: { in: taskIds } } })) as any
+    } else {
```
Comment (sazanrjb) [inline src/app/api/notification/notification.service.ts:358]: no need for else, just return

File: src/app/api/notification/notification.service.ts [hunk 3]
```diff
@@ -336,4 +347,16 @@ export class NotificationService extends BaseService {
...
     return { senderId, recipientId, recipientIds, actionUser, companyName }
   }
+
+  async getAllForTasks<T extends UserRole>(
+    tasks: Task[],
+  ): Promise<T extends UserRole.Client ? ClientNotification[] : InternalUserNotification[]> {
+    const taskIds = tasks.map((task) => task.id)
+
+    if (this.user.role === UserRole.Client) {
+      return (await this.db.clientNotification.findMany({ where: { taskId: { in: taskIds } } })) as any
```
Comment (sazanrjb) [inline src/app/api/notification/notification.service.ts:357]: we may not want to use any

File: src/app/api/notification/validate-count/validateCount.controller.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  const user = await authenticate(req)
+  const userId = z.string().safeParse(user.clientId)
+  if (userId.error) {
+    throw new APIError(httpStatus.UNAUTHORIZED, 'Only client users are allowed to access this feature')
+  }
+
+  const validateCountService = new ValidateCountService(user)
+  await validateCountService.fixClientNotificationCount(userId.data)
+
+  return NextResponse.json({ message: 'Validated counts successufully' })
```
Comment (sazanrjb) [inline src/app/api/notification/validate-count/validateCount.controller.ts:23]: `successfully`

**Review by sazanrjb — APPROVED** [review]

**Review by sazanrjb — APPROVED** [review]

**Review by arpandhakal — APPROVED** [review]

**Pull request merged**